components/api/src/main/java/com/hotels/styx/api/ByteStreamAggregator.java

```diff
@@ -0,0 +1,91 @@
+/*
+  Copyright (C) 2013-2018 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.api;
+
+import io.netty.buffer.CompositeByteBuf;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.netty.buffer.Unpooled.compositeBuffer;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+class ByteStreamAggregator implements Subscriber<Buffer> {
+    private final Publisher<Buffer> upstream;
+    private final int maxSize;
+    private final CompletableFuture<Buffer> future = new CompletableFuture<>();
+    private final AtomicBoolean active = new AtomicBoolean();
+    private final CompositeByteBuf aggregated = compositeBuffer();
+    private Subscription subscription;
+
+    public ByteStreamAggregator(Publisher<Buffer> upstream, int maxSize) {
+        this.upstream = requireNonNull(upstream);
+        this.maxSize = maxSize;
+    }
+
+    public CompletableFuture<Buffer> apply() {
```

Comment:
That would expose a half-constructed `ByteStreamAggregator` object to the `upstream`.

Also because the `ByteStreamAggregator` is package private, the `apply` method is never visible to Styx-API users, and therefore we don't have to solve this problem right now.